src/net_processing.cpp

```diff
@@ -4161,6 +4161,12 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             LogPrint(BCLog::MEMPOOLREJ, "%s from peer=%d was not accepted: %s\n", tx.GetHash().ToString(),
                 pfrom.GetId(),
                 state.ToString());
+            TRACE4(mempool, rejected,
```

Comment:
I initially suggested to @virtu to include the peer_id (and thus place it in net_processing.cpp). But based on the 'mempool' scope of this PR might be better to place this tracepoint somewhere transaction source (p2p/wallet/rpc,..) agnostic.

Reply:
My reasoning was to follow the precedent set by the debug log, which only records rejected transactions received by peers. I assumed this would be the most interesting case, especially since it allowed the collection of peer data (as suggested by @0xB10C). At the time this approach seemed pragmatic, but now it appears to clash with a clean design.

Let me know what you think of the following change:
1. Remove all networking-related information from the `mempool:rejected` tracepoint to narrow the scope the mempool tracepoints to mempool data, making for a clean interface.
2. Call the `mempool:rejected` tracepoint after [this loop](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1424-L1425) if `test_accept` is `false`.

Reply:
Sounds good. Might also make sense to call the tracepoint for the `test_accept == false` case. Maybe even pass `test_accept` as argument to allow filtering in e.g. BPF or downstream tools.

Reply:
Implemented in [069af24](https://github.com/bitcoin/bitcoin/commit/069af2404c0b71da2857766752d2296e621ca754).

This also simplified the functional tests, since validation no longer requires determining node address and port.

To test, for example, that transactions received via RPC are covered, one can try spending an already-spent tx via `bitcoin-cli`. Example (courtesy of [stackoverflow](https://stackoverflow.com/questions/38493893/heres-how-to-send-raw-transaction-btc-using-bitcoin-cli-command)):

```
bitcoin-cli sendrawtransaction 0100000001e34ac1e2baac09c366fce1c2245536bda8f7db0f6685862aecf53ebd69f9a89c000000006a47304402203e8a16522da80cef66bacfbc0c800c6d52c4a26d1d86a54e0a1b76d661f020c9022010397f00149f2a8fb2bc5bca52f2d7a7f87e3897a273ef54b277e4af52051a06012103c9700559f690c4a9182faa8bed88ad8a0c563777ac1d3f00fd44ea6c71dc5127ffffffff02a0252600000000001976a914d90d36e98f62968d2bc9bbd68107564a156a9bcf88ac50622500000000001976a91407bdb518fa2e6089fd810235cf1100c9c13d1fd288ac00000000
```